### ISSUE TITLE
com.thinkaurelius.titan.diskstorage.idmanagement.ConsistentKeyIDAuthority.getIDBlock()

### ISSUE THREAD

**Issue by zyoo** [thinkaurelius/titan#1174]
throw new TemporaryLockingException(String.format("Reached timeout %d (%s elapsed) when attempting to allocate id block on partition(%d)-namespace(%d)",timeout, methodTime.toString(), partition, idNamespace)); at ConsistenKeyIDAuthority.java(329)

Exception stack:
Caused by: java.util.IllegalFormatConversionException: d != com.thinkaurelius.titan.diskstorage.util.time.StandardDuration
        at java.util.Formatter$FormatSpecifier.failConversion(Formatter.java:4045)
        at java.util.Formatter$FormatSpecifier.printInteger(Formatter.java:2748)
        at java.util.Formatter$FormatSpecifier.print(Formatter.java:2702)
        at java.util.Formatter.format(Formatter.java:2488)
        at java.util.Formatter.format(Formatter.java:2423)
        at java.lang.String.format(String.java:2792)
        at com.thinkaurelius.titan.diskstorage.idmanagement.ConsistentKeyIDAuthority.getIDBlock(ConsistentKeyIDAuthority.java:329)
        at com.thinkaurelius.titan.graphdb.database.idassigner.StandardIDPool.renewBuffer(StandardIDPool.java:172)
        at com.thinkaurelius.titan.graphdb.database.idassigner.StandardIDPool.access$100(StandardIDPool.java:27)
        at com.thinkaurelius.titan.graphdb.database.idassigner.StandardIDPool$IDBlockRunnable.run(StandardIDPool.java:236)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)

thanks

**Comment by ghost** [comment]
Dan, any remarks or hints before I look into this ?

**Comment by ghost** [comment]
@zyoo :  For completeness, what version of Titan were you using when you saw this exception ?
Thanks,

**Comment by ghost** [comment]
Looking at the code and line numbers, I speculate this was generated on Titan 0.5.4

**Comment by ghost** [comment]
Pull request 1076 in Titan 1.0 
for Issue 1019 "Better Support For Timestamp"  found here:
https://github.com/thinkaurelius/titan/commit/f13611c5687999ab4162df30c9a3d5a8251d3f89#diff-0084690d5160735b1c14b7a7988a0e09

removed the StandardDuration custom class that appears in the above 0.5.4 stack trace here:

Caused by: java.util.IllegalFormatConversionException: d != com.thinkaurelius.titan.diskstorage.util.time.StandardDuration

dalaro marked this investigation for Titan 1.1:

> > dalaro added this to the Titan1.1 milestone 27 days ago

No longer looking at 0.5.4 code since that is back level and this is slated for 1.1 per Dan.

Now looking at Titan 1.1 code to see if I can tell if pull request 1076 addressed this problem in the process of making TimeStamp improvements, or if there is still potential for a similar problem to exist.

**Comment by ghost** [comment]
This small sample program may reproduce the problem described in the original stack trace.  This uses Titan 1.1 class types - so perhaps the problem still exists.  Will see if I can propose a fix:

```
package com.example;

import com.thinkaurelius.titan.diskstorage.util.time.Timer;
import com.thinkaurelius.titan.diskstorage.util.time.TimestampProvider;

import com.thinkaurelius.titan.diskstorage.locking.TemporaryLockingException;
import com.thinkaurelius.titan.diskstorage.util.time.TimestampProviders;


import java.time.Duration;
import java.time.Instant;


/**
 * This is an attempt to reproduce a problem mentioned in
 * Issue #1174 but in Titan 1.1 style because that is where the "fix"
 * is potentially listed to go.
 *
 */
public class Driver11
{
    public static void main(String[] args)
    {
        Driver11 me = new Driver11();
        me.doWork();
    }

    // here we take the line that was blowing up in Titan, and try to get it to blow
    // up with the refactoring that occurred for Titan 1.0 for this code.
    //
    // This code emulates the ConsistentKeyIDAuthority.java code in Titan, the getIDBlock() method
    // where the original exception reported in the defect was produced.
    private void doWork()
    {
        try
        {
            // It would appear that the default timer is MICRO which
            // according to comments in the code is really Milliseconds
            TimestampProvider times = TimestampProviders.MICRO;
            final Timer methodTime = times.getTimer().start();

            Instant firstInst = Instant.now();
            // randomly selected delay
            Thread.sleep(3010);
            Instant secondInst = Instant.now();
            // try a positive Duration first
            Duration timeout = Duration.between(firstInst, secondInst);

            //  FROM TITAN CODE: partition>=0 && partition<(1<<partitionBitWdith)
            final int partition = 2;

            //FROM TITAN CODE: can be any non-negative value
            final int idNamespace = 2;

            throw new TemporaryLockingException(String.format("Reached timeout %d (%s elapsed) when attempting to allocate id block on partition(%d)-namespace(%d)",
                    timeout, methodTime.toString(), partition, idNamespace));
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}
```

And the results:

```
java.util.IllegalFormatConversionException: d != java.time.Duration
    at java.util.Formatter$FormatSpecifier.failConversion(Formatter.java:4302)
    at java.util.Formatter$FormatSpecifier.printInteger(Formatter.java:2793)
    at java.util.Formatter$FormatSpecifier.print(Formatter.java:2747)
    at java.util.Formatter.format(Formatter.java:2520)
    at java.util.Formatter.format(Formatter.java:2455)
    at java.lang.String.format(String.java:2928)
    at com.ibm.example.Driver11.doWork(Driver11.java:55)
    at com.ibm.example.Driver11.main(Driver11.java:26)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
```

**Comment by ghost** [comment]
I don't know how to cause this problem to happen using Titan itself but the test program seems to uncover the issue.

It appears calling timeout.getNano() to convert the Duration to a long fixes the problem of formatting this with a %d.  Fix submitted in pull request: #1241

```
 throw new TemporaryLockingException(String.format("Reached timeout %d (%s elapsed) when attempting to allocate id block on partition(%d)-namespace(%d)",
 timeout.getNano(), methodTime.toString(), partition, idNamespace));
```

**Comment by spmallette** [comment]
merged #1241 - thanks
